contrib/golang/filters/http/source/golang_filter.h

```diff
@@ -0,0 +1,259 @@
+  bool doTrailer(ProcessorState& state, Http::HeaderMap& trailers);
+  bool doTrailerGo(ProcessorState& state, Http::HeaderMap& trailers);
+
+  uint64_t getMergedConfigId(ProcessorState& state);
+
+  void continueEncodeLocalReply(ProcessorState& state);
+  void continueStatusInternal(GolangStatus status);
+  void continueData(ProcessorState& state);
+
+  void onHeadersModified();
+
+  void sendLocalReplyInternal(Http::Code response_code, absl::string_view body_text,
+                              std::function<void(Http::ResponseHeaderMap& headers)> modify_headers,
+                              Grpc::Status::GrpcStatus grpc_status, absl::string_view details);
+
+  const FilterConfigSharedPtr config_;
+  Dso::DsoInstancePtr dynamic_lib_;
+
+  Http::RequestOrResponseHeaderMap* headers_ ABSL_GUARDED_BY(mutex_){nullptr};
+  Http::HeaderMap* trailers_ ABSL_GUARDED_BY(mutex_){nullptr};
+
+  // save temp values from local reply
+  Http::RequestOrResponseHeaderMap* local_headers_{nullptr};
+  Http::HeaderMap* local_trailers_{nullptr};
+
+  // The state of the filter on both the encoding and decoding side.
+  DecodingProcessorState decoding_state_;
+  EncodingProcessorState encoding_state_;
+
+  Grpc::Context& context_;
+  uint64_t cost_time_decode_{0};
+  uint64_t cost_time_encode_{0};
+  uint64_t cost_time_mem_{0};
+  uint64_t stream_id_{0};
+
+  httpRequestInternal* req_{nullptr};
+
+  // lock for has_destroyed_ and the functions get/set/copy/remove/etc that operate on the
+  // headers_/trailers_/etc, to avoid race between envoy c thread and go thread (when calling back
+  // from go). it should also be okay without this lock in most cases, just for extreme case.
+  Thread::MutexBasicLockable mutex_{};
+  bool has_destroyed_ ABSL_GUARDED_BY(mutex_){false};
+
+  // other filter trigger sendLocalReply during go processing in async.
+  // will wait go return before continue.
+  // this variable is read/write in safe thread, do no need lock.
+  bool local_reply_waiting_go_{false};
+
+  // the filter enter encoding phase
+  bool enter_encoding_{false};
+};
+
+// Go code only touch the fields in httpRequest
+struct httpRequestInternal : httpRequest {
+  std::weak_ptr<Filter> filter_;
+  // anchor a string temporarily, make sure it won't be freed before copied to Go.
+  std::string strValue;
+  httpRequestInternal(std::weak_ptr<Filter> f) { filter_ = f; }
+  std::weak_ptr<Filter> weakFilter() { return filter_; }
+};
+
+// used to count function execution time
+template <typename T = std::chrono::microseconds> struct Measure {
+  template <typename F, typename... Args> static typename T::rep execution(F func, Args&&... args) {
+    auto start = std::chrono::steady_clock::now(); // NO_CHECK_FORMAT(real_time)
+
+    func(std::forward<Args>(args)...);
+
+    auto duration = std::chrono::duration_cast<T>(
+        std::chrono::steady_clock::now() - // NO_CHECK_FORMAT(real_time)
+        start);
+
+    return duration.count();
+  }
+};
```

Comment:
Where is this used?

Reply:
> Where is this used?

@mattklein123 This is a generic function to count the time consumed by function calls, I will add cgo time consumption statistics in the next version. So the implementation of the function is kept for now.

Reply:
Done